src/main/kotlin/com/github/shirleh/datacollection/DataCollectionHandler.kt

```diff
@@ -154,4 +155,40 @@ object DataCollectionHandler {
 
         logger.exit()
     }
+
+    /**
+     * Collects voice data from the incoming [events].
+     */
+    suspend fun collectVoiceData(events: Flow<VoiceStateUpdateEvent>) = events.collect { saveVoiceData(it) }
+
+    private fun saveVoiceData(event: VoiceStateUpdateEvent) {
+        logger.entry(event)
+
+        val voiceState = event.current
+
+        val guildId = voiceState.guildId.asString()
+        val guildMemberId = voiceState.userId.asString()
+
+        val channelId = if (voiceState.channelId.isPresent) {
+            voiceState.channelId.get().asString()
+        } else {
+            event.old.get().channelId.get().asString()
+        }
```

Comment:
Ideally you should never have to use `isPresent` and `get` ([reference: recipe 4](https://blogs.oracle.com/javamagazine/12-recipes-for-using-the-optional-class-as-its-meant-to-be-used)). Instead, use the more idiomatic (and functional!) variant `Optional#map#orElse`! 

Line 175 can also go boom twice with NPE because of the two naive `get()`s :O 

So packaging it all up and it should look soooomething like this? (I did this without IDE :see_no_evil: )
```suggestion
        val channelId = voiceState.channelId
            .map { it.asString }
            .orElse { 
                event.old
                    .map { event -> event.channelId.map { it.asString }.orElse("") }
                    .orElse("") 
            }
```
It's super unwieldy though haha :joy: Perhaps there's a better way to handle nested `Optional`s :stuck_out_tongue: But it should give you the general gist of it! :blush:

Reply:
Hi,

This seems to work, I tried using flatMap to simplify it a bit.
```kotlin
voiceState.channelId
            .map { it.asString() }
            .orElse(
                event.old
                    .flatMap { old -> old.channelId.map { it.asString() } }
                    .orElse("")
            )
```